src/Sdk/WebApi/WebApi/BrokerHttpClient.cs

```diff
@@ -103,24 +103,31 @@ public async Task<TaskAgentMessage> GetRunnerMessageAsync(
                 new HttpMethod("GET"),
                 requestUri: requestUri,
                 queryParameters: queryParams,
+                readErrorBody: true,
                 cancellationToken: cancellationToken);
 
             if (result.IsSuccess)
             {
                 return result.Value;
             }
 
-            // the only time we throw a `Forbidden` exception from Listener /messages is when the runner is
-            // disable_update and is too old to poll
-            if (result.StatusCode == HttpStatusCode.Forbidden)
+            if (TryParseErrorBody(result.ErrorBody, out BrokerError brokerError))
             {
-                throw new AccessDeniedException($"{result.Error} Runner version v{runnerVersion} is deprecated and cannot receive messages.")
+                if (result.StatusCode == HttpStatusCode.Forbidden && brokerError.Code == 1)
                 {
-                    ErrorCode = 1
-                };
+                    throw new AccessDeniedException($"{result.Error} Runner version v{runnerVersion} is deprecated and cannot receive messages.")
+                    {
+                        ErrorCode = 1
+                    };
+                }
+            }
+
+            if (result.StatusCode == HttpStatusCode.Forbidden)
```

Comment:
Is this temporary back-compat logic? If so, can we comment for readers, e.g. [like this](https://github.com/actions/runner/blob/36c66c808363839ea7069bef87fd4572c8bb6c54/src/Sdk/RSWebApi/RunServiceHttpClient.cs#L110)

Reply:
Guessing in the future we rely on the message body to create a strongly typed exception (`AccessDeniedException` in this case). Otherwise we consider it to be a generic exception.

Reply:
updated